src.ts/providers/provider-websocket.ts

```diff
@@ -38,66 +40,134 @@ export type WebSocketCreator = () => WebSocketLike;
+                this.initWebSocket(primeWebsocket, '0x');
+                await this.waitShardReady('0x');
+                const shards = await this.getRunningLocations();
+                shards.forEach((shard) => {
+                    const port = 8200 + 20 * shard[0] + shard[1];
+                    const websocket = new _WebSocket(url.split(":")[0] + ":" + url.split(":")[1] + ":" + port) as WebSocketLike;
+                    this.initWebSocket(websocket, `0x${shard[0].toString(16)}${shard[1].toString(16)}`);
+                    this.#websocket.push(websocket);
+                    this._urlMap.set(`0x${shard[0].toString(16)}${shard[1].toString(16)}`, websocket );
+                });
+                // create array of all shards using flatMap
+                await Promise.all(shards.map((shard) => {
+                    this.waitShardReady(`0x${shard[0].toString(16)}${shard[1].toString(16)}`);
+                }))
+            }
+            return
+        } else if(typeof urls === 'function') {
+            // if WebSocketCreator
+            this._urlMap.set('0x', urls());
+            const primeWebsocket = urls();
+            this.#websocket.push(primeWebsocket);
+            this.initWebSocket(primeWebsocket, '0x')
+            await this.waitShardReady('0x');
+            const shards = await this.getRunningLocations();
+            shards.forEach((shard) => {
+                const port = 8200 + 20 * shard[0] + shard[1];
+                const websocket = new _WebSocket(this.#websocket[0].url.split(":")[0] + ":" + this.#websocket[0].url.split(":")[1] + ":" + port) as WebSocketLike
+                this.initWebSocket(websocket, `0x${shard[0].toString(16)}${shard[1].toString(16)}`)
+                this.#websocket.push(websocket);
+                this._urlMap.set(`0x${shard[0].toString(16)}${shard[1].toString(16)}`, websocket );
+            });
+            return;
+        } else {
+            // if WebSocketLike
+            const primeConnect = () => { return urls; };
+            this.initWebSocket(urls as WebSocketLike, '0x');
+            await this.waitShardReady('0x');
+            const primeWebSocket = primeConnect()
+            this._urlMap.set('0x', primeWebSocket as WebSocketLike);
+            this.#websocket.push(primeWebSocket as WebSocketLike);
+            const shards = await this.getRunningLocations();
+            shards.forEach((shard) => {
+                const port = 8200 + 20 * shard[0] + shard[1];
+                const websocket = new _WebSocket((urls as WebSocketLike).url.split(":")[0] + ":" + (urls as WebSocketLike).url.split(":")[1] + ":" + port) as WebSocketLike;
+                this.initWebSocket(websocket, `0x${shard[0].toString(16)}${shard[1].toString(16)}`);
+                this.#websocket.push(websocket);
+                this._urlMap.set(`0x${shard[0].toString(16)}${shard[1].toString(16)}`, websocket );
+            });
+            return;
+        }
```

Comment:
Looks like there some repeated code here. Lets refactor to only use the following logic once rather than in every condition

```typescript
const shards = await this.getRunningLocations();
shards.forEach((shard) => {
    const port = 8200 + 20 * shard[0] + shard[1];
    const websocket = new _WebSocket((urls as WebSocketLike).url.split(":")[0] + ":" + (urls as WebSocketLike).url.split(":")[1] + ":" + port) as WebSocketLike;
    this.initWebSocket(websocket, `0x${shard[0].toString(16)}${shard[1].toString(16)}`);
    this.#websocket.push(websocket);
    this._urlMap.set(`0x${shard[0].toString(16)}${shard[1].toString(16)}`, websocket );
});
return;
```